### ISSUE TITLE
Significant CPU usage when cluster is idle on large machine

### ISSUE THREAD

**Issue by pwaller** [dask/distributed#1313]
Hi, I'm running distributed on a single large machine - an ec2 m4.16xlarge, which has 64 hyperthreads.

I'm using it with dask. Distributed version is `1.18.0`, on Python 2.7.

I've had various performance issues, but the easiest thing to report is that when idle, `top` reports that python is consuming more than 100% CPU with a few percent spent in many different PIDs. This seems to be causing the (otherwise idle) system to context switch at a rate of ~70k / sec. Curiously, during actual execution the context switching rate drops to ~a few k.

I note that @mrocklin [reported something similar](https://github.com/dask/distributed/issues/531#issuecomment-247445278).

---

`perf top` on one of the processes shows a lot of time in `sys_futex` and `sys_select`.

<details>
<summary>Click to expand `perf top` sample</summary>

```
+   46.42%     0.10%  [kernel]                  [k] entry_SYSCALL_64_fastpath                                                                                                                    ▒
+   17.48%     0.05%  [kernel]                  [k] sys_futex                                                                                                                                    ▒
+   17.42%     0.41%  [kernel]                  [k] do_futex                                                                                                                                     ▒
+   16.35%     0.33%  libc-2.24.so              [.] __select                                                                                                                                     ▒
    15.56%     0.05%  [kernel]                  [k] sys_select                                                                                                                                   ▒
+   15.23%     0.16%  [kernel]                  [k] schedule                                                                                                                                     ▒
+   14.66%     0.68%  [kernel]                  [k] __schedule                                                                                                                                   ▒
+   12.95%     0.98%  [kernel]                  [k] core_sys_select                                                                                                                              ▒
+   11.98%     0.60%  [kernel]                  [k] do_select                                                                                                                                    ▒
+   11.33%     0.26%  [kernel]                  [k] poll_schedule_timeout                                                                                                                        ▒
+   10.77%     0.31%  [kernel]                  [k] schedule_hrtimeout_range_clock                                                                                                               ▒
+   10.74%    10.65%  python2.7                 [.] PyEval_EvalFrameEx                                                                                                                           ▒
+   10.13%     0.11%  libpthread-2.24.so        [.] do_futex_wait.constprop.3                                                                                                                    ▒
+    9.56%     0.88%  libpthread-2.24.so        [.] sem_post@@GLIBC_2.2.5                                                                                                                        ▒
+    8.83%     0.61%  [kernel]                  [k] futex_wait                                                                                                                                   ▒
+    7.67%     0.29%  [kernel]                  [k] futex_wake                                                                                                                                   ▒
+    7.53%     0.05%  [kernel]                  [k] futex_wait_queue_me                                                                                                                          ▒
+    6.39%     0.00%  [kernel]                  [k] wake_up_q                                                                                                                                    ▒
+    6.35%     3.33%  [kernel]                  [k] try_to_wake_up                                                                                                                               ▒
+    5.97%     0.09%  [kernel]                  [k] __perf_event_task_sched_out                                                                                                                  ▒
+    5.88%     0.17%  [kernel]                  [k] ctx_sched_out                                                                                                                                ▒
+    5.78%     5.78%  [kernel]                  [k] native_write_msr                                                                                                                             ▒
+    5.48%     0.15%  [kernel]                  [k] x86_pmu_disable                                                                                                                              ▒
     5.33%     0.00%  [kernel]                  [k] perf_pmu_disable.part.92                                                                                                                     ▒
+    4.02%     0.25%  [kernel]                  [k] proc_readfd_common                                                                                                                           ▒
+    4.01%     0.01%  [kernel]                  [k] deactivate_task                                                                                                                              ▒
     3.42%     0.00%  libpthread-2.24.so        [.] __waitpid                                                                                                                                    ▒
     3.40%     0.00%  [kernel]                  [k] sys_wait4                                                                                                                                    ▒
+    3.31%     0.54%  [kernel]                  [k] dequeue_entity                                                                                                                               ▒
     3.28%     2.20%  [kernel]                  [k] do_wait                                                                                                                                      ▒
     3.11%     0.09%  [kernel]                  [k] dequeue_task_fair                                                                                                                            ▒
+    3.09%     0.18%  [kernel]                  [k] proc_fill_cache                                                                                                                              ▒
+    2.77%     0.49%  [kernel]                  [k] finish_task_switch                                                                                                                           ▒
     2.74%     2.74%  [kernel]                  [k] pvclock_clocksource_read                                                                                                                     ▒
+    2.72%     0.16%  [kernel]                  [k] page_fault                                                                                                                                   ▒
     2.56%     0.00%  [kernel]                  [k] do_page_fault                                                                                                                                ▒
     2.53%     0.01%  [kernel]                  [k] __do_page_fault                                                                                                                              ▒
+    2.47%     0.65%  [vdso]                    [.] __vdso_gettimeofday                                                                                                                          ▒
     2.28%     0.05%  [kernel]                  [k] tick_program_event                                                                                                                           ▒
     2.28%     0.23%  [kernel]                  [k] handle_mm_fault                                                                                     
```

</details>

---

[`pyflame`](https://github.com/uber/pyflame) appears to show that the time is spent in [`distributed/process.py:_process_one_message:151`](
https://github.com/dask/distributed/blob/7985689ad26e02f90d16494581d9729f49d9bbf2/distributed/process.py#L151) (which is doing a `Queue.get`).

![image](https://user-images.githubusercontent.com/438648/28961122-b7468eda-78f0-11e7-8110-a266c05dc189.png)

---

In addition to the above ("easy to report") issue, big computations frequently spend 50% of the CPU time in the kernel (continuously for many minutes - something I expect seconds to take to run instead takes a minute and spends 1h of sys-CPU time across the 64 cores), seemingly doing very many page faults. If anyone has an idea of how to tell what is causing a minor page fault, I'd love to hear it. The behaviour is inconsistent. Sometimes, this doesn't happen and the 1s computation takes 1s. Sometimes it takes a minute.

I don't know if this latter issue is related to the high idle CPU usage, but the fact the machine is busy when it should be idle is making it harder to diagnose what is going wrong.

**Comment by pwaller** [comment]
Using [syscount](https://github.com/brendangregg/perf-tools/blob/5a511f5f775cfbc0569e6039435361cecd22dd86/syscount) I see 160,000 syscalls per second while idle, so 2,500 syscalls/process/second, if they are evenly distributed. Assuming all the python processes are doing the same thing, that's 50 syscalls per 0.02s wait on the idle spin-loop, so it looks like `Queue.get` with a wait is quite expensive in terms of causing a lot of activity at this scale.

**Comment by mrocklin** [comment]
Thank you for the excellent analysis @pwaller !  

I see two solutions here:

1.  Increase the timeout time up from 20ms to something larger, like 500ms.  This would decrease responsiveness but that's probably fine.
2.  Change nanny processes to watch several workers so that we can have one multiprocessing queue to watch that is shared among many workers.  This would keep responsiveness but is a larger change and would require more work.

Option 1, increasing the timeout time, is relatively trivial to do.  Ideally we would want to find a timeout duration that balanced responsiveness and CPU load.

**Comment by pwaller** [comment]
If it was 500ms, we should expect the context switches and syscalls/sec to drop by 25x (0.5 / 0.02), so that would bring context switches down to ~2 k/sec and syscalls down to ~6k/sec. CPU usage should go from ~100% to ~4%.

I'd prefer if there was a way to make the machine actually idle, since this is obviously still consuming energy. That's still a CPU day every month spent just on idling...

**Comment by mrocklin** [comment]
From my read on this code it's trying to do two things:

1.  read from a multiprocessing queue to get a stop or termination signal
2.  check the status of the process to see if it closed on its own

Currently it switches back and forth between these two by reading on the queue with a short timeout, checking the process, and then reading on the queue again.  There are other ways to handle concurrency, and probably other ways to solve this problem generally.

Is there a mechanism that you recommend?

**Comment by pwaller** [comment]
My python multi{thread,process}ing is a little rusty.

Is it possible to instead arrange that you wait on the process somewhere, and when the wait completes, the queue gets written into? That way there is no need for a spin wait.

The go programmer in me would do something like:

```go
select {
case msg := <-q:
... handle msg ...
case <-processExit:
... handle process exit
}
```

elsewhere:

```go
exitStatus := process.Wait()
processExit <- exitStatus
```

If it were not straightforward to do a simultaneous wait on two things you could communicate the exit status in-band on `q`:

```go
q <- ... message about exiting ...
```

But I don't know if there is some subtlety I'm unaware of in terms of who is actually waiting on the process.

For that matter, I can't find anything waiting on the process... is it in fact the join? OK, please forgive me thinking aloud as I write this - so in that case - how about doing the join in another thread - and when join returns the process has exited. So then signal `q`. Is there some reason this wouldn't work?

I see there is already [some in-band signalling going on with `q`](https://github.com/dask/distributed/blob/7985689ad26e02f90d16494581d9729f49d9bbf2/distributed/process.py#L130-L136).

**Comment by pwaller** [comment]
As another aside - another issue I have encountered is that occasionally during importing C modules (in jupyter), the kernel segfaults. It's clear that in this case lots of python processes are left over spinning CPU - so I guess in that case the distributed processes aren't being told to exit? Does this sound like a plausible bug? If so, it might be good to have some sort of involuntary mechanism to signal the distributed workers that the parent has exited, for example by having a pipe file descriptor, which when the read completes indicates that the parent process has exited (because the write side held by the parent process has been closed by the kernel during process shutdown).

**Comment by mrocklin** [comment]
cc @pitrou if he's around

**Comment by pwaller** [comment]
By the way, I'm continuing to do analysis and hit problems where stuff "just breaks" - sits there spinning.

And because of all these processes "doing work" it is quite hard to identify what has broken. I think this is a strong argument to make distributed idle properly, if at all possible.

**Comment by pitrou** [comment]
@pwaller, you're right it would be much better if things idled properly indeed. I did not envision that the timeout here would cause problems. This is compounded by the fact that under Python 2.7, `threading.Condition.wait`, and in turn `Queue.get`, uses a polling loop if a timeout is passed (i.e. things here should be significantly better on Python 3).

I think your suggestion is an interesting one. A secondary thread could wait for the process to end and then post a specific message on the queue.

**Comment by pitrou** [comment]
> another issue I have encountered is that occasionally during importing C modules (in jupyter), the (edit: ipython) kernel segfaults

Is that under Dask? Perhaps you may try and reproduce without Dask? It may be a Jupyter issue, or an issue with those C modules (what are they?). In any case, I recommend opening a separate issue.

**Comment by pwaller** [comment]
> Is that under Dask? Perhaps you may try and reproduce without Dask? It may be a Jupyter issue, or an issue with those C modules (what are they?). In any case, I recommend opening a separate issue.

It was various seemingly unrelated modules under different circumstances that I haven't been able to find a reproducer for yet. I don't believe it to be to do with dask specifically. I only mentioned it on this issue because (I believe) there was a distributed bug: when the jupyter kernel exited unexpectedly, the distributed workers are left running. So I when I went to continue my analysis by restarting, I had more and more copies of distributed workers running, all idling in the fashion described by this issue and consuming resources.

**Comment by mrocklin** [comment]
This may be partially resolved in #1317 .  We now have a separate thread that periodically checks the  process state and informs the queue when/if it dies.

**Comment by mrocklin** [comment]
That PR has a number of other changes.  These changes are restricted to the `distributed/process.py` file

**Comment by mrocklin** [comment]
@pwaller can you check if this is now resolved?

**Comment by pwaller** [comment]
I'm afraid I cannot. I am AFK for a bit, and the project where I was using the large machine has concluded for now (though there may be more in the future). I did participate in the code review though, and it certainly looks like things are moving in the right direction. Thanks for taking a look.

**Comment by mrocklin** [comment]
OK.  I appreciate your excellent issues over the past couple of weeks.  I regret that I wasn't very available to help engage here at the same time.  Hopefully we overlap better next time :)

Closing this as probably resolved.

**Comment by efire** [comment]
I am actually seeing the same issue with 1.22.0 (local cluster, python 2.7.6). Worker threads taking 10% CPU while cluster is idle.

**Comment by jakirkham** [comment]
Would recommend opening a new issue with more details. In particular please include an [MCVE]( https://stackoverflow.com/help/mcve ).
